File: Aula04-Memória/fibonacci.c
#include <stdio.h>
#include <time.h>

long fibonaccis[2] = {1, 1};
long fibonacci(int a);

int main(void)
{
   int y;
   printf("Posição do número desejado: ");
   scanf("%i", &y);
   clock_t begin = clock();
   printf("%lu\n", fibonacci(y));
   clock_t end = clock();
   printf("Tempo total: %f\n", (double)(end-begin)/ CLOCKS_PER_SEC);
}

long fibonacci(int a)
{
    if (a == 1 || a == 2)
    {
    	return 1;
    }
    else
    {
    	for (int i = 0; i <= a-2; i++)
    	{
    	    long tmp = fibonaccis[1];
    	    fibonaccis[1] = fibonaccis[0] + fibonaccis[1];
    	    fibonaccis[0] = tmp;
    	}
    	return fibonaccis[1];
    }
}

File: week_06/fibonacci.py
fibonaccis = dict()


def main():
    """

    :return: Fibonacci number
    """
    number = int(input("Posição do número: "))
    print(f"O {number}º número da sequência de Fibonacci é {fibonacci(number)}")


def fibonacci(n):
    if n == 1 or n == 2:
        return 1
    else:
        if n in fibonaccis.keys():
            return fibonaccis[n]
        else:
            fibonaccis[n] = fibonacci(n-1) + fibonacci(n-2)
            return fibonaccis[n]


if __name__ == '__main__':
    main()
